crates/sui-core/src/execution_cache.rs

```diff
@@ -61,7 +61,127 @@ impl ExecutionCacheMetrics {
+    pub testing_api: Arc<dyn TestingAPI>,
+}
+
+impl ExecutionCacheTraitPointers {
+    pub fn new<T>(cache: Arc<T>) -> Self
+    where
+        T: ExecutionCacheRead
+            + ExecutionCacheWrite
+            + BackingStore
+            + BackingPackageStore
+            + ObjectStore
+            + ExecutionCacheReconfigAPI
+            + AccumulatorStore
+            + CheckpointCache
+            + StateSyncAPI
+            + ExecutionCacheCommit
+            + TestingAPI
+            + 'static,
+    {
+        Self {
+            cache_reader: cache.clone(),
+            cache_writer: cache.clone(),
+            backing_store: cache.clone(),
+            backing_package_store: cache.clone(),
+            object_store: cache.clone(),
+            reconfig_api: cache.clone(),
+            accumulator_store: cache.clone(),
+            checkpoint_cache: cache.clone(),
+            state_sync_store: cache.clone(),
+            cache_commit: cache.clone(),
+            testing_api: cache.clone(),
+        }
+    }
+}
+
+static ENABLE_WRITEBACK_CACHE_ENV_VAR: &str = "ENABLE_WRITEBACK_CACHE";
+
+#[derive(Debug)]
+pub enum ExecutionCacheConfigType {
+    WritebackCache,
+    PassthroughCache,
+}
+
+pub fn choose_execution_cache(config: &ExecutionCacheConfig) -> ExecutionCacheConfigType {
+    #[cfg(msim)]
+    {
+        let mut use_random_cache = None;
+        sui_macros::fail_point_if!("select-random-cache", || {
+            let random = rand::random::<bool>();
+            tracing::info!("Randomly selecting cache: {}", random);
+            use_random_cache = Some(random);
+        });
+        if let Some(random) = use_random_cache {
+            if random {
+                return ExecutionCacheConfigType::PassthroughCache;
+            } else {
+                return ExecutionCacheConfigType::WritebackCache;
+            }
+        }
+    }
+
+    if std::env::var(ENABLE_WRITEBACK_CACHE_ENV_VAR).is_ok()
+        || matches!(config, ExecutionCacheConfig::WritebackCache { .. })
+    {
+        ExecutionCacheConfigType::WritebackCache
+    } else {
+        ExecutionCacheConfigType::PassthroughCache
+    }
+}
+
+pub fn build_execution_cache(
+    epoch_start_config: &EpochStartConfiguration,
+    prometheus_registry: &Registry,
+    store: &Arc<AuthorityStore>,
+) -> ExecutionCacheTraitPointers {
+    let execution_cache_metrics = Arc::new(ExecutionCacheMetrics::new(prometheus_registry));
+
+    let cache = epoch_start_config.execution_cache_type();
+    tracing::info!("using cache impl {:?}", cache);
+    match cache {
+        ExecutionCacheConfigType::WritebackCache => ExecutionCacheTraitPointers::new(
+            WritebackCache::new(store.clone(), execution_cache_metrics).into(),
+        ),
+        ExecutionCacheConfigType::PassthroughCache => ExecutionCacheTraitPointers::new(
+            PassthroughCache::new(store.clone(), execution_cache_metrics).into(),
+        ),
+    }
+}
+
+pub fn build_execution_cache_for_tests(
```

Comment:
Nit: there is no easy way to combine this function with above? The default `EpochStartConfiguration` also checks ENABLE_WRITEBACK_CACHE_ENV_VAR.

Reply:
hmm, i need to rename this function, its not even used by tests. let me think